rustls/src/lib.rs

```diff
@@ -311,6 +311,10 @@
 //!   Note that aws-lc-rs has additional build-time dependencies like cmake.
 //!   See [the documentation](https://aws.github.io/aws-lc-rs/requirements/index.html) for details.
 //!
+//! - `fips`: enable support for FIPS140-3-approved cryptography, via the aws-lc-rs crate.
+//!   This feature enables the `aws_lc_rs` feature, which makes the rustls crate depend
+//!   on [aws-lc-rs](https://github.com/aws/aws-lc-rs).
```

Comment:
I think we would do a breaking change where the `fips` feature activates `aws-lc-rs?/fips` and `ring?/fips` features (I am assuming that will be required for the additional build- and link-time complexity required), but does not activate the `dep:aws-lc-rs` feature and does not expose the implicit-provider builder functions (eg `ClientConfig::builder()`). That means users need to set features `[ "fips", "ring" | "aws_lc_rs" ]` and use the explicit builder functions (`ClientConfig::builder_with_provider()`).

"Global features" would make this easier, I think, but it's unclear if they will ever come to fruition.

Reply:
It's not clear that we would necessarily get to do a breaking change? At least it's possible in theory that FIPS support in *ring* does not need semver-breaking changes (the addition of a `fips` Cargo feature or a `fips()` canary function certainly wouldn't be semver-breaking). In that case, enabling `fips` and `ring` for rustls would still select `aws-lc-rs` even though that would be highly surprising behavior.

Reply:
> At least it's possible in theory that FIPS support in _ring_ does not need semver-breaking changes (the addition of a `fips` Cargo feature or a `fips()` canary function certainly wouldn't be semver-breaking)

Yeah I accept that is theoretically possible, but equally that wouldn't be operating in FIPS mode. That means we're back to "I requested the `fips` feature but got defaults that aren't FIPS-approved" situation. The other theoretical possibility is that FIPS becomes the default for *ring*, in a minor release -- that would break existing applications at runtime when rustls tries to use chacha20 (for example).

> That's at odds with Cargo's additivity guidelines but IMO it might be a better solution than this.

For me that would be a worse solution (because it requires complete and instant cooperation from everyone to turn off rustls's default features), unless we also drop `ring` from the default feature set. And, aside from the cooperation requirement, the usability around turning off a default feature is highly neglected (rust-lang/cargo/issues/11467, rust-lang/cargo/issues/3126).

The other options I see are: splitting up the crates (mentioned in #1184) or removing APIs that re-state the crate default features (opposite of #1744, but maybe #1642 is relevant).

Reply:
> Yeah I accept that is theoretically possible, but equally that wouldn't be operating in FIPS mode. That means we're back to "I requested the `fips` feature but got defaults that aren't FIPS-approved" situation. The other theoretical possibility is that FIPS becomes the default for _ring_, in a minor release -- that would break existing applications at runtime when rustls tries to use chacha20 (for example).

Okay, so since we're fairly sure that *ring* must make breaking changes to enable FIPS mode (because this removes some cryptographic algorithms) we will therefore have to make explicit changes in rustls to enable support for that version of ring, and can change our support for the `fips` flag to forward to `ring/fips` instead of `aws-lc-rs/fips`.

I can live with that.